### ISSUE TITLE
flake8 called via tox fails with FileNotFoundError and InvocationError

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#748]
In GitLab by @flowolf on Jun 2, 2020, 07:26

*Please describe how you installed Flake8*

Example:

```
$ pip install flake8
and via tox
$ tox -e flake8 # tox runs the install
```

Tox version:
```
$ tox --version
3.15.1 imported from /home/flo/sync/git/yessssms/venv_yessssms/lib/python3.6/site-packages/tox/__init__.py

```

*Please provide the exact, unmodified output of `flake8 --bug-report`*
```{
  "dependencies": [],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.6.8",
    "system": "Linux"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "flake8-colors",
      "version": "0.1.0"
    },
    {
      "is_local": false,
      "plugin": "flake8-docstrings",
      "version": "1.5.0, pydocstyle: 5.0.2"
    },
    {
      "is_local": false,
      "plugin": "flake8_typing_imports",
      "version": "1.9.0"
    },
    {
      "is_local": false,
      "plugin": "import-order",
      "version": "0.18.1"
    },
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "naming",
      "version": "0.10.0"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.6.0"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "2.2.0"
    }
  ],
  "version": "3.8.2"
}
```
*Please describe the problem or feature*

I'm running flake8 via tox since version 3.8.0 the build pipeline fails because flake8 does not find the files to be tested.
version 3.7.9 was the last version that worked.

*If this is a bug report, please explain with examples (and example code) what you expected to happen and what actually happened.*

Here are some examples:

running **flake8 (current 3.8.2)** via tox fails:

```
$ cat tox.ini
# tox (https://tox.readthedocs.io/) is a tool for running tests

[tox]
envlist = py35, py36, py37, py38, flake8, pylint

[testenv]
deps =
    pytest-cov
    requests-mock
    pytest
commands =
    pytest
    pytest -v --cov=YesssSMS tests
    pytest --cov=YesssSMS --cov-report=html tests
# cat YesssSMS/YesssSMS.py,cover

[testenv:flake8]
basepython = python3
ignore = E203, E266, E501, W503, F403, F401
max-line-length = 88
max-complexity = 18
select = B,C,E,F,W,T4,B9
skip_install = true
changedir = {toxinidir}
deps =
    flake8
    flake8-colors
    flake8-docstrings>=1.3.1
    flake8-import-order>=0.9
    flake8-typing-imports>=1.1
    pep8-naming
commands =
    flake8 YesssSMS/*py tests/*py setup.py

[testenv:pylint]
deps = pylint
commands =
    pylint -d C0330,C0103 YesssSMS
$ tox -e flake8
flake8 recreate: /home/flo/sync/git/yessssms/.tox/flake8
flake8 installdeps: flake8, flake8-colors, flake8-docstrings>=1.3.1, flake8-import-order>=0.9, flake8-typing-imports>=1.1, pep8-naming
flake8 installed: flake8==3.8.2,flake8-colors==0.1.6,flake8-docstrings==1.5.0,flake8-import-order==0.18.1,flake8-polyfill==1.0.2,flake8-typing-imports==1.9.0,importlib-metadata==1.6.0,mccabe==0.6.1,pep8-naming==0.10.0,pycodestyle==2.6.0,pydocstyle==5.0.2,pyflakes==2.2.0,snowballstemmer==2.0.0,YesssSMS==0.4.2,zipp==3.1.0
flake8 run-test-pre: PYTHONHASHSEED='2454754583'
flake8 run-test: commands[0] | flake8 'YesssSMS/*py' 'tests/*py' setup.py
YesssSMS/*py:0:1: E902 FileNotFoundError: [Errno 2] No such file or directory: 'YesssSMS/*py'
tests/*py:0:1: E902 FileNotFoundError: [Errno 2] No such file or directory: 'tests/*py'
ERROR: InvocationError for command /home/flo/sync/git/yessssms/.tox/flake8/bin/flake8 'YesssSMS/*py' 'tests/*py' setup.py (exited with code 1)
________________________________________________________________________________ summary ________________________________________________________________________________
ERROR:   flake8: commands failed
tox -e flake8  5,28s user 0,53s system 91% cpu 6,338 total

```

running tests via tox - does not work (**flake 3.8.0**)
```
$ cat tox.ini
# tox (https://tox.readthedocs.io/) is a tool for running tests

[tox]
envlist = py35, py36, py37, py38, flake8, pylint

[testenv]
deps =
    pytest-cov
    requests-mock
    pytest
commands =
    pytest
    pytest -v --cov=YesssSMS tests
    pytest --cov=YesssSMS --cov-report=html tests
# cat YesssSMS/YesssSMS.py,cover

[testenv:flake8]
basepython = python3
ignore = E203, E266, E501, W503, F403, F401
max-line-length = 88
max-complexity = 18
select = B,C,E,F,W,T4,B9
skip_install = true
changedir = {toxinidir}
deps =
    flake8==3.8.0
    flake8-colors
    flake8-docstrings>=1.3.1
    flake8-import-order>=0.9
    flake8-typing-imports>=1.1
    pep8-naming
commands =
    flake8 YesssSMS/*py tests/*py setup.py

[testenv:pylint]
deps = pylint
commands =
    pylint -d C0330,C0103 YesssSMS
$ tox -e flake8
flake8 installed: flake8==3.8.0,flake8-colors==0.1.6,flake8-docstrings==1.5.0,flake8-import-order==0.18.1,flake8-polyfill==1.0.2,flake8-typing-imports==1.9.0,importlib-metadata==1.6.0,mccabe==0.6.1,pep8-naming==0.10.0,pycodestyle==2.6.0,pydocstyle==5.0.2,pyflakes==2.2.0,snowballstemmer==2.0.0,YesssSMS==0.4.2,zipp==3.1.0
flake8 run-test-pre: PYTHONHASHSEED='4113297879'
flake8 run-test: commands[0] | flake8 'YesssSMS/*py' 'tests/*py' setup.py
YesssSMS/*py:0:1: E902 FileNotFoundError: [Errno 2] No such file or directory: 'YesssSMS/*py'
tests/*py:0:1: E902 FileNotFoundError: [Errno 2] No such file or directory: 'tests/*py'
ERROR: InvocationError for command /home/flo/sync/git/yessssms/.tox/flake8/bin/flake8 'YesssSMS/*py' 'tests/*py' setup.py (exited with code 1)
________________________________________________________________________________ summary ________________________________________________________________________________
ERROR:   flake8: commands failed
$                                                                                                                                                                    :( 
```

**version 3.7.9 works:**
```
$ cat tox.ini
# tox (https://tox.readthedocs.io/) is a tool for running tests

[tox]
envlist = py35, py36, py37, py38, flake8, pylint

[testenv]
deps =
    pytest-cov
    requests-mock
    pytest
commands =
    pytest
    pytest -v --cov=YesssSMS tests
    pytest --cov=YesssSMS --cov-report=html tests
# cat YesssSMS/YesssSMS.py,cover

[testenv:flake8]
basepython = python3
ignore = E203, E266, E501, W503, F403, F401
max-line-length = 88
max-complexity = 18
select = B,C,E,F,W,T4,B9
skip_install = true
changedir = {toxinidir}
deps =
    flake8==3.7.9
    flake8-colors
    flake8-docstrings>=1.3.1
    flake8-import-order>=0.9
    flake8-typing-imports>=1.1
    pep8-naming
commands =
    flake8 YesssSMS/*py tests/*py setup.py

[testenv:pylint]
deps = pylint
commands =
    pylint -d C0330,C0103 YesssSMS
$ tox -e flake8
flake8 recreate: /home/flo/sync/git/yessssms/.tox/flake8
cat tox.	flake8 installdeps: flake8==3.7.9, flake8-colors, flake8-docstrings>=1.3.1, flake8-import-order>=0.9, flake8-typing-imports>=1.1, pep8-naming
flake8 installed: entrypoints==0.3,flake8==3.7.9,flake8-colors==0.1.6,flake8-docstrings==1.5.0,flake8-import-order==0.18.1,flake8-polyfill==1.0.2,flake8-typing-imports==1.9.0,importlib-metadata==1.6.0,mccabe==0.6.1,pep8-naming==0.10.0,pycodestyle==2.5.0,pydocstyle==5.0.2,pyflakes==2.1.1,snowballstemmer==2.0.0,YesssSMS==0.4.2,zipp==3.1.0
flake8 run-test-pre: PYTHONHASHSEED='2523145459'
flake8 run-test: commands[0] | flake8 'YesssSMS/*py' 'tests/*py' setup.py
________________________________________________________________________________ summary ________________________________________________________________________________
  flake8: commands succeeded
  congratulations :)
tox -e flake8  5,16s user 0,54s system 91% cpu 6,223 total
$
```

running via command line, not via tox, works fine.
```
$ flake8 --version
3.8.2 (mccabe: 0.6.1, pycodestyle: 2.6.0, pyflakes: 2.2.0) CPython 3.6.8 on
Linux
$ flake8 YesssSMS/*py tests/*py setup.py
$ flake8 -v YesssSMS/*py tests/*py setup.py
flake8.plugins.manager    MainProcess     80 INFO     Loading entry-points for "flake8.extension".
flake8.plugins.manager    MainProcess    115 INFO     Loading entry-points for "flake8.report".
flake8.plugins.manager    MainProcess    150 INFO     Loading plugin "C90" from entry-point.
flake8.plugins.manager    MainProcess    152 INFO     Loading plugin "F" from entry-point.
flake8.plugins.manager    MainProcess    166 INFO     Loading plugin "pycodestyle.ambiguous_identifier" from entry-point.
flake8.plugins.manager    MainProcess    172 INFO     Loading plugin "pycodestyle.bare_except" from entry-point.
flake8.plugins.manager    MainProcess    172 INFO     Loading plugin "pycodestyle.blank_lines" from entry-point.
flake8.plugins.manager    MainProcess    172 INFO     Loading plugin "pycodestyle.break_after_binary_operator" from entry-point.
flake8.plugins.manager    MainProcess    172 INFO     Loading plugin "pycodestyle.break_before_binary_operator" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.comparison_negative" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.comparison_to_singleton" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.comparison_type" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.compound_statements" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.continued_indentation" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.explicit_line_join" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.extraneous_whitespace" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.imports_on_separate_lines" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.indentation" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.maximum_doc_length" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.maximum_line_length" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.missing_whitespace" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.missing_whitespace_after_import_keyword" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.missing_whitespace_around_operator" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.module_imports_on_top_of_file" from entry-point.
flake8.plugins.manager    MainProcess    173 INFO     Loading plugin "pycodestyle.python_3000_async_await_keywords" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.python_3000_backticks" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.python_3000_has_key" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.python_3000_invalid_escape_sequence" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.python_3000_not_equal" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.python_3000_raise_comma" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.tabs_obsolete" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.tabs_or_spaces" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.trailing_blank_lines" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.trailing_whitespace" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.whitespace_around_comma" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.whitespace_around_keywords" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.whitespace_around_named_parameter_equals" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.whitespace_around_operator" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.whitespace_before_comment" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "pycodestyle.whitespace_before_parameters" from entry-point.
flake8.plugins.manager    MainProcess    174 INFO     Loading plugin "default" from entry-point.
flake8.plugins.manager    MainProcess    175 INFO     Loading plugin "pylint" from entry-point.
flake8.plugins.manager    MainProcess    175 INFO     Loading plugin "quiet-filename" from entry-point.
flake8.plugins.manager    MainProcess    175 INFO     Loading plugin "quiet-nothing" from entry-point.
flake8.options.manager    MainProcess    175 WARNING  option --max-complexity: please update from optparse string `type=` to argparse callable `type=` -- this will be an error in the future
flake8.checker            MainProcess    181 INFO     Making checkers
flake8.checker            MainProcess    191 INFO     Checking 10 files
flake8.main.application   MainProcess    603 INFO     Finished running
flake8.main.application   MainProcess    604 INFO     Reporting errors
flake8.main.application   MainProcess    604 INFO     Found a total of 5 violations and reported 0
$                                                                                                                                                                     
```

I tried reducing the config complexity of the tox call - did not work - 3.7.9 works - 3.8.2 fails:
```
$ $ tox -e flake8
GLOB sdist-make: /home/flo/sync/git/yessssms/setup.py
flake8 inst-nodeps: /home/flo/sync/git/yessssms/.tox/.tmp/package/1/YesssSMS-0.4.2.zip
flake8 installed: certifi==2020.4.5.1,chardet==3.0.4,flake8==3.8.2,idna==2.9,importlib-metadata==1.6.0,mccabe==0.6.1,pycodestyle==2.6.0,pyflakes==2.2.0,requests==2.23.0,urllib3==1.25.9,YesssSMS==0.4.2,zipp==3.1.0
flake8 run-test-pre: PYTHONHASHSEED='978424085'
flake8 run-test: commands[0] | flake8 --bug-report
{
  "dependencies": [],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.6.8",
    "system": "Linux"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.6.0"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "2.2.0"
    }
  ],
  "version": "3.8.2"
}
flake8 run-test: commands[1] | flake8 'YesssSMS/*py' 'tests/*py' setup.py
YesssSMS/*py:0:1: E902 FileNotFoundError: [Errno 2] No such file or directory: 'YesssSMS/*py'
tests/*py:0:1: E902 FileNotFoundError: [Errno 2] No such file or directory: 'tests/*py'
ERROR: InvocationError for command /home/flo/sync/git/yessssms/.tox/flake8/bin/flake8 'YesssSMS/*py' 'tests/*py' setup.py (exited with code 1)
________________________________________________________________________________ summary ________________________________________________________________________________
ERROR:   flake8: commands failed

$ cat tox.ini
# tox (https://tox.readthedocs.io/) is a tool for running tests

[tox]
envlist = py35, py36, py37, py38, flake8, pylint

[testenv]
deps =
    pytest-cov
    requests-mock
    pytest
commands =
    pytest
    pytest -v --cov=YesssSMS tests
    pytest --cov=YesssSMS --cov-report=html tests
# cat YesssSMS/YesssSMS.py,cover

[testenv:flake8]
basepython = python3
#ignore = E203, E266, E501, W503, F403, F401
#max-line-length = 88
#max-complexity = 18
#select = B,C,E,F,W,T4,B9
#skip_install = true
#changedir = {toxinidir}
deps =
    flake8 --bug-report
    flake8
#    flake8-colors
#    flake8-docstrings>=1.3.1
#    flake8-import-order>=0.9
#    flake8-typing-imports>=1.1
#    pep8-naming
commands =
    flake8 YesssSMS/*py tests/*py setup.py

[testenv:pylint]
deps = pylint
commands =
    pylint -d C0330,C0103 YesssSMS
$ vi tox.ini
$ cat tox.ini
# tox (https://tox.readthedocs.io/) is a tool for running tests

[tox]
envlist = py35, py36, py37, py38, flake8, pylint

[testenv]
deps =
    pytest-cov
    requests-mock
    pytest
commands =
    pytest
    pytest -v --cov=YesssSMS tests
    pytest --cov=YesssSMS --cov-report=html tests
# cat YesssSMS/YesssSMS.py,cover

[testenv:flake8]
basepython = python3
#ignore = E203, E266, E501, W503, F403, F401
#max-line-length = 88
#max-complexity = 18
#select = B,C,E,F,W,T4,B9
#skip_install = true
#changedir = {toxinidir}
deps =
    flake8==3.7.9
#    flake8-colors
#    flake8-docstrings>=1.3.1
#    flake8-import-order>=0.9
#    flake8-typing-imports>=1.1
#    pep8-naming
commands =
    flake8 YesssSMS/*py tests/*py setup.py

[testenv:pylint]
deps = pylint
commands =
    pylint -d C0330,C0103 YesssSMS
$ tox -e flake8
GLOB sdist-make: /home/flo/sync/git/yessssms/setup.py
flake8 recreate: /home/flo/sync/git/yessssms/.tox/flake8
flake8 installdeps: flake8==3.7.9
flake8 inst: /home/flo/sync/git/yessssms/.tox/.tmp/package/1/YesssSMS-0.4.2.zip
flake8 installed: certifi==2020.4.5.1,chardet==3.0.4,entrypoints==0.3,flake8==3.7.9,idna==2.9,mccabe==0.6.1,pycodestyle==2.5.0,pyflakes==2.1.1,requests==2.23.0,urllib3==1.25.9,YesssSMS==0.4.2
flake8 run-test-pre: PYTHONHASHSEED='4178242403'
flake8 run-test: commands[0] | flake8 'YesssSMS/*py' 'tests/*py' setup.py
________________________________________________________________________________ summary ________________________________________________________________________________
  flake8: commands succeeded
  congratulations :)
tox -e flake8  7,10s user 0,85s system 92% cpu 8,603 total


```

You can have a look at the failing build of the pipeline on my project site: https://gitlab.com/flowolf/yessssms/-/jobs/555889702

Thanks for your help.

**Comment by asottile** [comment]
In GitLab by @asottile on Jun 2, 2020, 08:11

globs (`foo/*.py`) are a shell feature, `tox` does not invoke a shell so they are not expanded.  since they are not expanded they are looking for _literal_ files with stars in them.

replace your command with `flake8 YesssSMS/ tests/ setup.py` and it will work

it used to "work" in flake8 3.7.x because flake8 would **silently ignore not found files** (and so your check was silently only validating `setup.py` -- definitely not what you wanted!)

(disclaimer: I'm also a tox maintainer)

**Comment by asottile** [comment]
In GitLab by @asottile on Jun 2, 2020, 08:11

closed

**Comment by asottile** [comment]
In GitLab by @flowolf on Jun 2, 2020, 08:56

you are absolutely right. that's not what I intended. ;)
good thing it fails now with this wrong call. fixed that - I got some stuff to clean up.

Thanks for your help!

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jun 2, 2020, 15:26

@asottile this is a clear regression. 2.x and 3.x up until 3.8 expanded those globs. We did it explicitly in the code. I suspect something Eric did around configuration.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jun 2, 2020, 15:26

reopened

**Comment by asottile** [comment]
In GitLab by @asottile on Jun 2, 2020, 15:33

not that I can tell @sigmavirus24 

```console
$ flake8 '*.py'
$ cat t.py 
import os
```

and here's the flake8 2.x behaviour:

```
$ flake8 '*.py'
/tmp/x/venv/lib/python3.8/site-packages/pycodestyle.py:111: FutureWarning: Possible nested set at position 1
  EXTRANEOUS_WHITESPACE_REGEX = re.compile(r'[[({] | []}),;:]')
*.py:1:1: E902 FileNotFoundError: [Errno 2] No such file or directory: '*.py'
```

**Comment by asottile** [comment]
In GitLab by @asottile on Jun 2, 2020, 15:42

closed

**Comment by asottile** [comment]
In GitLab by @asottile on Jun 2, 2020, 15:43

to be clear, the change that made these apparent is I fixed E902 to re-trigger in 3.8 in !404 (nice coincidence on the PR number lol!)
